src/mpid/ch4/shm/posix/eager/iqueue/iqueue_send.h

```diff
@@ -0,0 +1,136 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+#ifndef POSIX_EAGER_IQUEUE_SEND_H_INCLUDED
+#define POSIX_EAGER_IQUEUE_SEND_H_INCLUDED
+
+#include "iqueue_impl.h"
+
+MPL_STATIC_INLINE_PREFIX MPIDI_POSIX_EAGER_IQUEUE_cell_t
+    * MPIDI_POSIX_EAGER_IQUEUE_new_cell(MPIDI_POSIX_EAGER_IQUEUE_transport_t * transport)
+{
+    int i;
+    for (i = 0; i < transport->num_cells; i++) {
+        MPIDI_POSIX_EAGER_IQUEUE_cell_t *cell = MPIDI_POSIX_EAGER_IQUEUE_THIS_CELL(transport, i);
+        if (cell->type == MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_NULL) {
+            return cell;
+        }
+    }
+    return NULL;
```

Comment:
Every time we have to search available cell from the head of the list, can we use the buffer pool structure to mange it (e.g., like request object) ?

Reply:
It is a main feature. To free a cell only need to just set one flag. Synchronization doesn't needed. The total number of used cells always as small as possible (no extra cache pollution).

Reply:
I think a detailed documentation is needed to explain the iqueue design.

Reply:
@SergeyGubanov - I think writing such detailed documentation goes beyond what I'm going to be able to produce given that I didn't write any of this code. Can you provide a document about the design (I can help format and/or figure out where to put it if you can come up with the content).